src/init.cpp

```diff
@@ -797,6 +797,19 @@ ServiceFlags nLocalServices = NODE_NETWORK;
 
 }
 
+[[noreturn]] static void new_handler_terminate()
+{
+    // Rather than throwing std::bad-alloc if allocation fails, terminate
+    // immediately to (try to) avoid chain corruption.
+    // Since LogPrintf may itself allocate memory, set the handler directly
+    // to terminate first.
+    std::set_new_handler(std::terminate);
+    LogPrintf("Error: Out of memory. Terminating.\n");
```

Comment:
I don't think this is threadsafe (ie, a concurrent OOM could cause it to terminate in the middle of a LogPrintf or print two LogPrintfs). I would make the handler have some `switch(atomic counter++)`. Then it is both re-entrant and threadsafe. (note that set_new_handler is threadsafe itself, just is threadsafe in this usage)

I think this sort of implementation should be ok:

```c++
static std::thread::id first_entrant{std::this_thread::get_id()};
static bool first_pass {true};
if (first_entrant == std::this_thread::get_id()) {
    if (first_pass) {
        first_pass = false;
        // Single faulted
        std::terminate();
    } else {
        // Double faulted
    }
} else {
    sleep();
}
```


Either of the above should be safe for concurrent OOM as well as double OOM faults.